To vectorize long datatype as mask index

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92057
* #91755
* __->__ #91076

In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

EikanWang committed:
To vectorize long datatype as mask index

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2185,7 +2185,11 @@ def get_grads(args):
         call_forwards_backwards(f)
         orig_grad = get_grads(args)
-        self.assertEqual(orig_grad, compiled_grad)
+        grad_to_check_device = orig_grad[0] if isinstance(orig_grad, list) else orig_grad
+        is_cpu_device = isinstance(grad_to_check_device, torch.Tensor) and grad_to_check_device.device == torch.device("cpu")
+        atol = 2e-05 if is_cpu_device else 1.3e-05
+        rtol = 1.3e-05 if is_cpu_device else 1.3e-06
+        self.assertEqual(orig_grad, compiled_grad, atol=atol, rtol=rtol)
```
This is irrelevant to the changes of tis PR?

EikanWang replied:
This PR vectorized the `masked`, hence it will impact the precision.

jgong5 replied:
weird, vectorized `masked` is just vectorizing the condition checks, that alters the compute result too?

EikanWang replied:
Originally, the loop body was not vectorized due to `masked`. But the loop body would be vectorized w/ this PR. Therefore, it could impact the compute result.

desertfire replied:
Agree with @Chillee , this looks like an aot-autograd only test.

EikanWang replied:
The failed case is `test_aot_autograd_exhaustive_svd_lowrank_cpu_float32 `- https://github.com/pytorch/pytorch/actions/runs/3826715657/jobs/6510779512#step:10:8623

I encountered the accuracy issue w/ this feature enabled. Therefore, I refined the test case for CPU-only. But it is strange that I could not reproduce the issue w/ the latest PyTorch code. @Chillee , @desertfire

EikanWang replied:
@desertfire , @Chillee , I will reuse the original code. Are you okay with that?

Chillee replied:
@EikanWang yeah I'm fine with reverting this change.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -999,15 +1128,59 @@ def reduction(name, dtype, src_dtype, reduction_type, index, value):
-                self.simd_vec = False
+                current_node: torch.fx.Node = V.interpreter.current_node
+
+                loop_range = {}
+                assert len(self.ranges) == len(self.itervars)
+                for idx in range(len(self.ranges)):
+                    loop_range[self.itervars[idx]] = self.ranges[idx]
+                expr_val = sympy.simplify(sympy_subs(expr, loop_range))
```
You mean the range contains min and max while the expression could not calculate the range?

jgong5 replied:
I mean simply plugging in the range of individual loop variables into the expression might not yield the right range of the expression. A naïve example: `100-i0` where `i0` ranges from 0 to 30. The range of `100-i0` is 70-100. You cannot plug in 30 to `100-i0`.

EikanWang replied:
Got your point.

This checker happens before the loop transformation, and then the loop interval is always `[0, range)` now. Hence, I think we do not need to take this situation into account.

jgong5 replied:
But here you assume that the index expr gets to its max when the loop variable gets to the max, which is not always true (like the example I provided).

EikanWang replied:
Fixed.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -285,17 +316,19 @@ def reciprocal(a):
 
     @staticmethod
     def constant(val, dtype):
+        proposed_dtype = V.interpreter.current_node.meta["dtype"]
```
The `dtype` should always be there. But assert also makes sense here. I will add it.

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -370,6 +403,40 @@ def expm1(x):
+        assert dtype == torch.int64
+        assert V.interpreter.current_node.meta["dtype"] == torch.int32
+        assert V.interpreter.current_node.meta["most_inner_loop_irrevelant"]
+        return f"at::vec::Vectorized<int>(static_cast<int>({cexpr(V.kernel.rename_indexing(expr))}))"
```
Got your point.

This checker happens before the loop transformation, and then the loop interval is always `[0, range)` now. Hence, I think we do not need to take this situation into account.

jgong5 replied:
Sure, let's do it in a separate PR.

EikanWang commented on torch/_inductor/ir.py:
```diff
@@ -4110,7 +4115,12 @@ def __call__(self):
+        interpreter_shim = InterpreterShim(graph, submodules)
+        V.set_interpreter_handler(interpreter_shim)
+        res = interpreter_shim.run(V.get_ops_handler())
+        V.set_interpreter_handler(_orig_interpreter)
```
Fixed.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -370,6 +403,40 @@ def expm1(x):
+    @staticmethod
+    def index_expr(expr, dtype):
+        assert dtype == torch.int64
+        assert V.interpreter.current_node.meta["dtype"] == torch.int32
+        assert V.interpreter.current_node.meta["most_inner_loop_irrevelant"]
```
Fixed.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -370,6 +403,40 @@ def expm1(x):
 
+    @staticmethod
+    def masked(mask, body, other):
+        assert V.interpreter.current_node.meta["is_masked_load"]
```
Done.

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -231,6 +234,34 @@ def erf(x):
+
+    @staticmethod
+    def and_(x, y):
+        return f"{x} & {y}"
```
Rebased

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +69,26 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
+
+#if defined(CPU_CAPABILITY_AVX512) || defined(CPU_CAPABILITY_AVX2)
+template <typename SRC>
+inline at::vec::Vectorized<float> to_float_mask(at::vec::Vectorized<SRC>& src) {
```
Should this be a method defined on `at::vec::Vectorized` ?

EikanWang replied:
It is more specific for the `mask` semantic because the `src` always represents `True` or `False`. I'm sort of concerned it is not general enough to add it to `at::vec::Vectorized` as the `at::vec` is a common interface for multiple datatypes and both x86 and ARM. May I know your suggestion?

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1043,15 +1188,83 @@ def reduction(name, dtype, src_dtype, reduction_type, index, value):
                 return is_supported_dtype
 
             @staticmethod
             def index_expr(expr, dtype):
```
The logic is a bit complicated. Give a brief description on what is supported in the docstring?

EikanWang replied:
I submitted another PR on top of this ghstack. Do you think it could address your comment?

jgong5 replied:
Sounds good.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -397,6 +407,43 @@ def expm1(x):
+        elif isinstance(other, float):
+            code.writeline(f"auto {var} = at::vec::Vectorized<float>({other});")
+        else:
+            code.writeline(f"auto {var} = at::vec::Vectorized<float>({other!r});")
```
May I know why do we use the conversion flag "!r" here?

EikanWang replied:
"!r" means `repr` - https://docs.python.org/3/library/functions.html#repr 
I copied it from the non-vec version.

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -962,21 +1040,30 @@ def could_vec(self, name: str, index: sympy.Expr):
-        if not V.graph.get_dtype(name) in [torch.float, torch.float32]:
+        store_dtype = V.graph.get_dtype(name)
+
+        current_node: torch.fx.Node = V.interpreter.current_node
+        current_node.meta["dtype"] = store_dtype
```
I submitted another PR on top of this ghstack. Do you think it could address your comment?

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -398,6 +408,43 @@ def expm1(x):
+                f"auto {var} = at::vec::Vectorized<float>(std::numeric_limits<float>::infinity());"
+            )
+        elif isinstance(other, float):
+            code.writeline(f"auto {var} = at::vec::Vectorized<float>({other});")
```
Won't the next case do the same thing for a float?

EikanWang replied:
Fixed by removing this logic.

jansel commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +69,26 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
+  return at::vec::Vectorized<float>(_mm512_cvtepi32_ps(src));
+#endif
+}
+#endif
```
newline end of file

EikanWang replied:
Fixed.

desertfire commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1197,6 +1278,52 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+    def is_supported_cmp(self, node: torch.fx.Node):
+        def get_node_dtype(node):
+            if type(node) == torch.fx.Node:
+                return None if "dtype" not in node.meta else node.meta["dtype"]
```
Nit: you can use `node.meta.get("dtype", None)`

desertfire commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1197,6 +1278,52 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+
+        left_dtype, right_dtype = get_cmp_dtypes(node)
+        if left_dtype is None or right_dtype is None:
+            # TODO(Eikan): Should be conservative?
```
Shouldn't we start with being conservative and leave the aggressive option as TODO?

EikanWang replied:
Actually, it is a missing piece to record, deduce and propagate the data type of every expression. Hence, we could not get the real data type of the left expression or the right expression. I refined the comment. By the way, I'm working on the data type analysis.

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@jansel @desertfire , may I know if I have addressed your comments?

desertfire reviewed (approved):
Looks good to me, but please address Horace's and my comments to why `test/functorch/test_aotdispatch.py` is relevant for this PR before your merge.

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "To vectorize long datatype as mask index"


In this PR, we record the current fx node being executed to cache additional information to simply the vectorization checker. In addition, we supported `masked` in this PR by simplifying it as `mask_load` to support `max_pool2d`.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge
